Sources/SwiftBSON/ExtendedJSONDecoder.swift

```diff
@@ -28,21 +56,138 @@ public class ExtendedJSONDecoder {
     /// - Returns: Decoded representation of the JSON input as an instance of `T`.
     /// - Throws: `DecodingError` if the JSON data is corrupt or if any value throws an error during decoding.
     public func decode<T: Decodable>(_: T.Type, from data: Data) throws -> T {
-        // Data --> JSON --> BSON --> T
+        // Data --> JSONValue --> BSON --> T
         // Takes in JSON as `Data` encoded with `.utf8` and runs it through a `JSONDecoder` to get an
         // instance of the `JSON` enum.
-
-        // In earlier versions of Swift, JSONDecoder doesn't support decoding "fragments" at the top level, so we wrap
-        // the data in an array to guarantee it always decodes properly.
-        let wrappedData = "[".utf8 + data + "]".utf8
-        let json = try JSONDecoder().decode([JSON].self, from: wrappedData)[0]
+        let json = try JSONParser().parse(bytes: data)
 
         // Then a `BSON` enum instance is created via the `JSON`.
-        let bson = try BSON(fromExtJSON: json, keyPath: [])
+        // let bson = try json.toBSON(keyPath: [])
+        let bson = try self.decodeBSONFromJSON(json, keyPath: [])
+        // let bson = BSON(fromExtJSON: json)
 
         // The `BSON` is then passed through a `BSONDecoder` where it is outputted as a `T`
         let bsonDecoder = BSONDecoder()
         bsonDecoder.userInfo = self.userInfo
         return try bsonDecoder.decode(T.self, fromBSON: bson)
     }
+
+    /// Decode a `BSON` from the given extended JSON.
```

Comment:
the methods here enable us to recursively build up the document without having to allocate many `ByteBuffers` for each subdocument, which is much faster. The caveat is that it is a lot less natural to do so. For scalars, we can operate pretty similarly to before, so we return those as `BSON`. (technically we should build up code with scope similarly, but I opted to treat it as a scalar for simplicity). For non-scalars (arrays and documents), we return them from `decodeScalar` as-is and then do further processing here in `decodeBSONFromJSON` and recursive calls in descending from `appendElement`.